benchmarks/perftest_decoding.cpp

```diff
@@ -0,0 +1,82 @@
+#include "spdlog/spdlog.h"
+
+#include "index_types.hpp"
+#include "memory_source.hpp"
+#include "util/do_not_optimize_away.hpp"
+
+#include <iostream>
+
+using pisa::do_not_optimize_away;
+using pisa::get_time_usecs;
+
+template <typename IndexType, bool decode_freqs>
+double perftest(IndexType const &index)
+{
+    auto start = get_time_usecs();
+    for (int i = 0; i < index.size(); i++) {
+        auto plist = index[i];
+        // Reads elements of each posting list.
+        for (size_t i = 0; i < plist.size(); ++i) {
+            // Puts pointer on next posting.
+            plist.next();
+
+            // Loads docid and freq (if required) of current posting. On the other hand,
+            // do_not_optimize_away() is used to avoid compiler optimizations. Otherwise,
+            // the compiler may decide not to execute the following sentences because the
+            // result isn't used.
+            do_not_optimize_away(plist.docid());
+            if (decode_freqs) {
```

Comment:
I think what you wanted is:

```cpp
if constexpr (decode_freqs) {
```

This will simply remove this branch for the version of the function with `decode_freqs=false`, so we won't be measuring the branch decision itself.

I don't expect this branch to affect the results in any noticeable way (after all, the loop itself has conditions as well and we're ok with it), but the only reason I see for passing `decode_freqs` as a template parameter is if you want to evaluate it in constexpr context. Otherwise this could simply be a function argument.